zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -474,132 +531,39 @@ impl DbFormatChange {
             return Ok(());
         };
 
-        // Note commitment tree de-duplication database upgrade task.
-
-        let version_for_pruning_trees =
-            Version::parse("25.1.1").expect("Hardcoded version string should be valid.");
-
-        // Check if we need to prune the note commitment trees in the database.
-        if older_disk_version < &version_for_pruning_trees {
-            let timer = CodeTimer::start();
-
-            // Prune duplicate Sapling note commitment trees.
-
-            // The last tree we checked.
-            let mut last_tree = db
-                .sapling_tree_by_height(&Height(0))
-                .expect("Checked above that the genesis block is in the database.");
-
-            // Run through all the possible duplicate trees in the finalized chain.
-            // The block after genesis is the first possible duplicate.
-            for (height, tree) in db.sapling_tree_by_height_range(Height(1)..=initial_tip_height) {
-                // Return early if there is a cancel signal.
-                if !matches!(cancel_receiver.try_recv(), Err(TryRecvError::Empty)) {
-                    return Err(CancelFormatChange);
-                }
-
-                // Delete any duplicate trees.
-                if tree == last_tree {
-                    let mut batch = DiskWriteBatch::new();
-                    batch.delete_sapling_tree(db, &height);
-                    db.write_batch(batch)
-                        .expect("Deleting Sapling note commitment trees should always succeed.");
-                }
-
-                // Compare against the last tree to find unique trees.
-                last_tree = tree;
+        // Apply or validate format upgrades
+        for upgrade in format_upgrades() {
+            if older_disk_version >= &upgrade.version() {
+                upgrade
+                    .validate(db, cancel_receiver)?
+                    .expect("failed to validate db format");
+                continue;
```

Comment:
I'm a bit confused about this. The original code does not seem to do validate if the disk version >= upgrade version. Is this required or just good to have? What is the performance impact?

Reply:
This change was wrong, I thought some of the upgrades were being validated here even when they weren't being applied, but none are.

> What is the performance impact?

It shouldn't have a significant performance impact, it won't block anything. These will still always run at startup anyway but should run after any required format upgrades have been applied.

Reply:
I'm assuming you're still working on this then?

Reply:
Removed in https://github.com/ZcashFoundation/zebra/pull/9263/commits/55b36028eb45e0349f56f7ebb24088f1f6c53e4e (by accepting a `min_version` in `format_upgrades()` and filtering out older upgrades altogether).